cts/cts-exec.in

```diff
@@ -102,6 +102,27 @@ def output_from_command(command):
     return output.split("\n")
 
 
+def killall(process_names=[], terminate=False):
+    """ Kill all instances of every process in a list """
+
+    if not process_names:
+        return
+
+    procs = []
+    for proc in psutil.process_iter(["name"]):
+        if proc.info["name"] in process_names:
+            procs.append(proc)
+
+    if terminate:
+        for proc in procs:
+            proc.terminate()
+        gone, alive = psutil.wait_procs(procs, timeout=3)
+        procs = alive
+
+    for proc in procs:
+        proc.kill()
```

Comment:
Minimizing TOCTOU risk is more important than efficiency when killing processes. You wouldn't think that there's enough time here for one of the processes to exit and a new one to reuse its pid, but it's just a matter of time before it surprises you. :) Better to check for and kill them one by one, even grabbing the entire process list again each time.

Of course there's also a chance that a relevant process will start after you check, but that's a less severe consequence and extremely unlikely if someone's running CTS tests.

Reply:
We should be good.

> Note: the way this class is bound to a process is uniquely via its PID. That means that if the process terminates and the OS reuses its PID you may end up interacting with another process. The only exceptions for which process identity is preemptively checked (via PID + creation time) is for the following methods: ... [kill()](https://psutil.readthedocs.io/en/latest/#psutil.Process.kill).

https://psutil.readthedocs.io/en/latest/#psutil.Process

> Kill the current process by using SIGKILL signal preemptively checking whether PID has been reused. 

https://psutil.readthedocs.io/en/latest/#psutil.Process.kill

The actual reuse check:
https://github.com/giampaolo/psutil/blob/379598312f60bf414afa8bf549f7f26af9e578ea/psutil/__init__.py#L571-L592
https://github.com/giampaolo/psutil/blob/379598312f60bf414afa8bf549f7f26af9e578ea/psutil/__init__.py#L266-L279

Reply:
BTW (not directly relevant but as a note) I pretty much grabbed the "kill several processes" pattern from here:
- https://psutil.readthedocs.io/en/latest/#psutil.wait_procs

Reply:
Ah, that's great! They actually put some thought into that one :)